### ISSUE TITLE
TRAM 

### ISSUE THREAD

**Issue by cclementi** [markovmodel/PyEMMA#825]
Hi, I am running TRAM with 500 states obtained by clustering the first 10 TICs from a total of 300,000 frames) and I get the following warnings that I don't understand:

finding connected set:   5% ( 29/500) [########                                                                                                                                           ] eta 01:55 -/home/mi/cclementi/anaconda2/lib/python2.7/site-packages/pyEMMA-2.2-py2.7-linux-x86_64.egg/pyemma/thermo/estimators/TRAM_estimator.py:212: EmptyState: Thermodynamic state 0 contains no samples after reducing to the connected set.
- ' contains no samples after reducing to the connected set.', EmptyState)
  /home/mi/cclementi/anaconda2/lib/python2.7/site-packages/pyEMMA-2.2-py2.7-linux-x86_64.egg/pyemma/thermo/estimators/TRAM_estimator.py:212: EmptyState: Thermodynamic state 1 contains no samples after reducing to the connected set.
- ' contains no samples after reducing to the connected set.', EmptyState)
  /home/mi/cclementi/anaconda2/lib/python2.7/site-packages/pyEMMA-2.2-py2.7-linux-x86_64.egg/pyemma/thermo/estimators/TRAM_estimator.py:212: EmptyState: Thermodynamic state 2 contains no samples after reducing to the connected set.
- ' contains no samples after reducing to the connected set.', EmptyState)
  /home/mi/cclementi/anaconda2/lib/python2.7/site-packages/pyEMMA-2.2-py2.7-linux-x86_64.egg/pyemma/thermo/estimators/TRAM_estimator.py:212: EmptyState: Thermodynamic state 3 contains no samples after reducing to the connected set.
- ' contains no samples after reducing to the connected set.', EmptyState)
  /home/mi/cclementi/anaconda2/lib/python2.7/site-packages/pyEMMA-2.2-py2.7-linux-x86_64.egg/pyemma/thermo/estimators/TRAM_estimator.py:235: EmptyState: Thermodynamic state 0contains no transitions after reducing to the connected set.
- 'contains no transitions after reducing to the connected set.', EmptyState)
  /home/mi/cclementi/anaconda2/lib/python2.7/site-packages/pyEMMA-2.2-py2.7-linux-x86_64.egg/pyemma/thermo/estimators/TRAM_estimator.py:235: EmptyState: Thermodynamic state 1contains no transitions after reducing to the connected set.
- 'contains no transitions after reducing to the connected set.', EmptyState)
  /home/mi/cclementi/anaconda2/lib/python2.7/site-packages/pyEMMA-2.2-py2.7-linux-x86_64.egg/pyemma/thermo/estimators/TRAM_estimator.py:235: EmptyState: Thermodynamic state 2contains no transitions after reducing to the connected set.
- 'contains no transitions after reducing to the connected set.', EmptyState)
  /home/mi/cclementi/anaconda2/lib/python2.7/site-packages/pyEMMA-2.2-py2.7-linux-x86_64.egg/pyemma/thermo/estimators/TRAM_estimator.py:235: EmptyState: Thermodynamic state 3contains no transitions after reducing to the connected set.
- 'contains no transitions after reducing to the connected set.', EmptyState)
  TRAM increment=3.1e-03/5.0e-14:   0% (     48/1500000) [                                                                                                                                            ] |/home/mi/cclementi/anaconda2/lib/python2.7/site-packages/pyEMMA-2.2-py2.7-linux-x86_64.egg/pyemma/thermo/estimators/TRAM_estimator.py:269: RuntimeWarning: divide by zero encountered in log
  N_dtram_accelerations=self.N_dtram_accelerations)
  TRAM increment=4.7e-04/5.0e-14:   0% (   1376/1500000) [  

It looks like there are two problems, the first one related to finding empty states, and the second one related to a division by zero. If I run with 100 states the first problem seem to disappear but the division by zero is still there. I am running with the options: 
connectivity= 'post_hoc_RE'
direct_space=True
N_dtram_acc=5
I can provide the full input file and data set if you like.

Thanks!

**Comment by fabian-paul** [comment]
Hi @cclementi,

The EmptyState state warnings are intentional. This means that the connectivity measure  'post_hoc_RE' sees no overlap between thermodynamic states {0,1,2} and the rest. I guess because the set of other thermodynamic states {3,4,..} is larger (in terms of total number of micro states), it got picked up as the largest connected set. This might actually not be what you want (?) But there seems to be an overlap problem with the data anyway.

The divide by zero warning is issued when we compute np.log(0) and is not an error. This is ok, because we are allowing infinite free energies. I just don't know how to silence this numpy warning.

**Comment by franknoe** [comment]
I think the Warning should not be shown, because it looks like something 
is seriously wrong.
1. Would this work?
   https://docs.python.org/2/library/warnings.html#temporarily-suppressing-warnings
2. Alternative: Only do np.log() on the nonzero index set, and set the 
   rest to Inf.

> Hi @cclementi https://github.com/cclementi,
> 
> The EmptyState state warnings are intentional. This means that the 
> connectivity measure 'post_hoc_RE' sees no overlap between 
> thermodynamic states {0,1,2} and the rest. I guess because the set of 
> other thermodynamic states {3,4,..} is larger (in terms of total 
> number of micro states), it got picked up as the largest connected 
> set. This might actually not be what you want (?) But there seems to 
> be an overlap problem with the data anyway.
> 
> The divide by zero warning is issued when we compute np.log(0) and is 
> not an error. This is ok, because we are allowing infinite free 
> energies. I just don't know how to silence this numpy warning.
> 
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub 
> https://github.com/markovmodel/PyEMMA/issues/825#issuecomment-223798086, 
> or mute the thread 
> https://github.com/notifications/unsubscribe/AGMeQtriMvbNyrLZ8G4aQQ4pw3gLbB3eks5qIn0UgaJpZM4IuMO9.

## 

---

Prof. Dr. Frank Noe
Head of Computational Molecular Biology group
Freie Universitaet Berlin

Phone: (+49) (0)30 838 75354
Web:   research.franknoe.de

## Mail:  Arnimallee 6, 14195 Berlin, Germany

**Comment by cclementi** [comment]
Thanks!

**Comment by fabian-paul** [comment]
@franknoe actually I don't like any of your suggestions. :-/
1. The divide by zero warning is a RuntimeWarning so it's very unspecific and we would have to catch the warning based on the error message string (which I don't like). 
2. Writing a fully numpy-compatible function with indexing and the generic support for scalar, lists, tuples, ... argument types that one expects from numpy is tough. 
So I would rather define:

``` Python
def dumb_log(x):
   with np.errstate(divide='ignore'):
       return np.log(x)
```

What's your opinion @cwehmeyer ?

**Comment by franknoe** [comment]
I'm happy with any other solution too, my only point is to not show a 
warning if the user hasn't done anything wrong.

Am 06/06/16 um 11:16 schrieb fabian-paul:

> @franknoe https://github.com/franknoe actually I don't like any of 
> your suggestions. :-/
> 1. The divide by zero warning is a RuntimeWarning so it's very 
> unspecific and we would have to catch the warning based on the error 
> message string (which I don't like).
> 2. Writing a fully numpy-compatible function with indexing and the 
> generic support for scalar, lists, tuples, ... argument types that one 
> expects from numpy is tough.
> So I would rather define:
> 
> def  dumb_log(x):
>     with  np.errstate(divide='ignore'):
>         return  np.log(x)
> 
> What's your opinion @cwehmeyer https://github.com/cwehmeyer ?
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub 
> https://github.com/markovmodel/PyEMMA/issues/825#issuecomment-223907512, 
> or mute the thread 
> https://github.com/notifications/unsubscribe/AGMeQnDTHqNbEGhtm-7IYGJuHljEFik-ks5qI-VtgaJpZM4IuMO9.

## 

---

Prof. Dr. Frank Noe
Head of Computational Molecular Biology group
Freie Universitaet Berlin

Phone: (+49) (0)30 838 75354
Web:   research.franknoe.de

## Mail:  Arnimallee 6, 14195 Berlin, Germany

**Comment by fabian-paul** [comment]
The division by zero warning is fixed by markovmodel/thermotools@67028efd70a4ffc7053cebf69a75433629dc2aca
As the rest of this issue is related to the sampling situation, I'm closing this here.
